components/post_view/embedded_bindings/button_binding/button_binding.tsx

```diff
@@ -0,0 +1,127 @@
+import {Channel} from 'mattermost-redux/types/channels';
+import {Post} from 'mattermost-redux/types/posts';
+
+import Markdown from 'components/markdown';
+import LoadingWrapper from 'components/widgets/loading/loading_wrapper';
+import {createCallContext, createCallRequest} from 'utils/apps';
+
+type Props = {
+    intl: IntlShape;
+    binding: AppBinding;
+    post: Post;
+    actions: {
+        doAppCall: (call: AppCallRequest, type: AppCallType, intl: IntlShape) => Promise<ActionResult>;
+        getChannel: (channelId: string) => Promise<ActionResult>;
+    };
+    sendEphemeralPost: (message: string, channelID?: string, rootID?: string) => void;
+}
+
+type State = {
+    executing: boolean;
+}
+
+export class ButtonBinding extends React.PureComponent<Props, State> {
+    constructor(props: Props) {
+        super(props);
+        this.state = {
+            executing: false,
+        };
+    }
+
+    handleClick = async () => {
+        const {binding, post} = this.props;
+        if (!binding.call) {
+            return;
+        }
+
+        let teamID = '';
+        const {data} = await this.props.actions.getChannel(post.channel_id) as {data?: any; error?: any};
+        if (data) {
+            const channel = data as Channel;
+            teamID = channel.team_id;
+        }
+
+        const context = createCallContext(
+            binding.app_id,
+            AppBindingLocations.IN_POST + '/' + binding.location,
+            post.channel_id,
+            teamID,
+            post.id,
+            post.root_id,
+        );
+        const call = createCallRequest(
+            binding.call,
+            context,
+            {post: AppExpandLevels.EXPAND_ALL},
+        );
+        this.setState({executing: true});
+        const res = await this.props.actions.doAppCall(call, AppCallTypes.SUBMIT, this.props.intl);
+
+        this.setState({executing: false});
+        const callResp = (res as {data: AppCallResponse}).data;
+        const ephemeral = (message: string) => this.props.sendEphemeralPost(message, this.props.post.channel_id, this.props.post.root_id);
+        switch (callResp.type) {
+        case AppCallResponseTypes.OK:
+            if (callResp.markdown) {
+                ephemeral(callResp.markdown);
+            }
+            break;
+        case AppCallResponseTypes.ERROR: {
+            const errorMessage = callResp.error || this.props.intl.formatMessage({id: 'apps.error.unknown', defaultMessage: 'Unknown error happenned'});
+            ephemeral(errorMessage);
+            break;
+        }
+        case AppCallResponseTypes.NAVIGATE:
+        case AppCallResponseTypes.FORM:
+            break;
+        default: {
+            const errorMessage = this.props.intl.formatMessage(
+                {id: 'apps.error.responses.unknown_type', defaultMessage: 'App response type not supported. Response type: {type}.'},
```

Comment:
The type may come from the remote server. Then, you are right. If someone were to add any markdown in the type, they could post it here. We could surround it by backticks to make it a "in-line code", but it could still use a backtick to close the "in-line code" and still put some markdown inside. Do we have already any markdown escape function that we could use here to avoid this?

Reply:
No, I don't think we have any escaping function for Markdown. I suppose we could use a regex like `/[\*~_\[\]\(\)]/g` and add a backslash before any of those characters, but I'm not sure if that's even necessary since I don't think it's any different than a websocket posting arbitrary Markdown anyway.

Reply:
My biggest fear is that an "evil app" could potentially do something like:
- Set the type as `unknown. Click [here](downloadviruslink) to solve the problem`
- The user would see: `App response type not supported. Response type: unknown. Click [here](downloadviruslink) to solve the problem.`
- The user would click that.

But if the "evil app" has access to do that, they could just return OK and put that information in the markdown field, so a message like that (or more convincing) is sent to the user.

So you are right, probably handling this is not necessary.

Reply:
Oh yeah, there's definitely an opportunity for that. I meant that it's no worse than what other types of integrations can do currently 😛